[hipDNN] Add MatmulNode to hipDNN frontend

sshlyapn opened the pull request:
## Motivation

Add Matmul node to the frontend and SDK graph (see ALMIOPEN-812 for the details)
Should be merged after https://github.com/ROCm/rocm-libraries/pull/3523

## Technical Details

- Added new `MatmulNode` implementation
- Added `Graph::matmul` method and exposed it as Python binding
- Added SDK schema and generated headers
- Minor: added missed `hipdnn_sdk_DIR` to support SDK discovery in custom Python builds

## Test Plan

- New unit tests: `projects/hipdnn/frontend/tests/TestMatmulNode.cpp` and `projects/hipdnn/frontend/tests/TestGraph.cpp`
- New Python sample: `projects/hipdnn/python/hipdnn_frontend/samples/matmul.py`

## Test Result

All new Matmul node tests pass

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

SamuelReeder commented:
Thanks for your contribution!

FYI, we just migrated the remaining `hipdnn_sdk` utilities to the `hipdnn_data_sdk` as part of our SDK partitioning. That is the source of the merge conflicts. Hopefully they're trivial fixes, but sorry for the inconvenience!

sshlyapn commented:
> Thanks for your contribution!
> 
> FYI, we just migrated the remaining `hipdnn_sdk` utilities to the `hipdnn_data_sdk` as part of our SDK partitioning. That is the source of the merge conflicts. Hopefully they're trivial fixes, but sorry for the inconvenience!

Got it, thanks for explaining! I've fixed the conflicts and updated the branch - should be all set now

SamuelReeder commented on projects/hipdnn/frontend/tests/TestMatmulNode.cpp:
```diff
@@ -0,0 +1,242 @@
+using namespace hipdnn_frontend;
+using namespace hipdnn_frontend::graph;
+
+TEST(TestMatmulNode, PreValidateNode_Succeeds2D)
```
I think we generally avoid underscores in test names because GTest reserves them for keywords and they can generate class name collisions. See: https://google.github.io/googletest/faq.html#why-should-test-suite-names-and-test-names-not-contain-underscore

sshlyapn replied:
Applied, thanks!

SamuelReeder commented on projects/hipdnn/frontend/include/hipdnn_frontend/node/MatmulNode.hpp:
```diff
@@ -0,0 +1,157 @@
+                                      + "batch dimensions for broadcasting at index "
+                                      + std::to_string(i) + ": A has dim=" + std::to_string(aDimVal)
+                                      + ", B has dim=" + std::to_string(bDimVal));
+        }
```
I believe instead of verifying divisibility of batch dimensions, we want to verify that either the dimensions are equal or one of them is 1. NumPy and cuDNN follow this standard:
- NumPy: https://numpy.org/doc/stable/user/basics.broadcasting.html#general-broadcasting-rules
- cuDNN: https://docs.nvidia.com/deeplearning/cudnn/backend/v9.17.1/api/cudnn-graph-library.html#cudnn-backend-operation-matmul-descriptor

PyTorch also follows NumPy's semantics.

Without this, the broadcast is ambiguous and the strides can't be represented by a constant (at least for this generic matmul node).

sshlyapn replied:
Yeah, I also initially thought about such broadcasting semantics; however, cuDNN allows more flexible [broadcasting rules](https://docs.nvidia.com/deeplearning/cudnn/backend/v9.17.1/api/cudnn-graph-library.html#cudnn-backend-operation-matmul-descriptor):

Case | Matrix A | Matrix B | Matrix C
-- | -- | -- | --
Single Matmul | 1 x M x K | 1 x K x N | 1 x M x N
Batch Matmul | B x M x K | B x K x N | B x M x N
Broadcast A | (B/c) x M x K | B x K x N | B x M x N
Broadcast B | B x M x K | (B/c) x K x N | B x M x N

Where: B indicates the batch size, c is a constant integer and a factor of B

So, if we would like alignment, we probably need to support such broadcasting as well

SamuelReeder replied:
This excerpt from that page may also be relevant:

> The functionality of having multiple batch dimensions allows you to have layouts where the batch is not packed at a single stride. This case is especially seen in multihead attention. c is only allowed to be B (leading to a batch dimension for 1) for matmul and matmul fusions. The other possible values of c are supported for Grouped Query Attention in the cuDNN Fused Flash Attention. [[source](https://docs.nvidia.com/deeplearning/cudnn/backend/v9.17.1/api/cudnn-graph-library.html#id101)]

It seems for matmul broadcasting, cuDNN also restricts one of the tensor dimensions to 1. The table they give is a little misleading, but maybe it is referring to Grouped Query Attention for the case where both dimensions can be inequal and greater than 1, and they have deterministic rules for broadcasting in that routine.

Unless I'm interpreting this incorrectly?

SamuelReeder replied:
Perhaps that only applies to tensors with 3 batch dimensions, but it's odd that it has different constraints.

SamuelReeder replied:
I think we can accept the more lenient broadcasting compatibility and let the plugin decide.

SamuelReeder commented on projects/hipdnn/frontend/include/hipdnn_frontend/node/MatmulNode.hpp:
```diff
@@ -0,0 +1,157 @@
+            b, ErrorCode::ATTRIBUTE_NOT_SET, "MatmulNode missing B input for pre-validation");
+
+        HIPDNN_RETURN_IF_FALSE(
+            c, ErrorCode::ATTRIBUTE_NOT_SET, "MatmulNode missing C output for pre-validation");
```
Nit: we could put this into a private helper and call in both methods to reduce duplication.

sshlyapn replied:
Moved to the helper functions, thanks!

SamuelReeder commented:
FYI I'm investigating the CI failure.

---

Edit: should be fixed.

BrianHarrisonAMD commented on projects/hipdnn/python/hipdnn_frontend/samples/matmul.py:
I don't think we need to add this with this PR, but we typically add C++ samples as well, and docs updates when adding new operation support.

This can come later once we have engine support for matmul though.

andrei-kochin replied:
@sshlyapn to follow-up
